crates/red_knot/src/module.rs

```diff
@@ -388,12 +389,70 @@ pub fn file_to_module(db: &dyn SemanticDb, file: FileId) -> QueryResult<Option<M
 //////////////////////////////////////////////////////
 
 /// Changes the module search paths to `search_paths`.
-pub fn set_module_search_paths(db: &mut dyn SemanticDb, search_paths: Vec<ModuleSearchPath>) {
+pub fn set_module_search_paths(db: &mut dyn SemanticDb, search_paths: ResolvedSearchPathOrder) {
     let jar: &mut SemanticJar = db.jar_mut();
 
     jar.module_resolver = ModuleResolver::new(search_paths);
 }
 
+const TYPESHED_STDLIB_DIRECTORY: &str = "stdlib";
+
+/// A resolved module resolution order, implementing PEP 561
+/// (with some small, deliberate differences)
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ResolvedSearchPathOrder(Vec<ModuleSearchPath>);
+
+impl Deref for ResolvedSearchPathOrder {
+    type Target = [ModuleSearchPath];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Implementation of the module resolution order from PEP-561.
+///
+/// - `extra_paths` is a list of user-provided paths
+///   that should take first priority in the module resolution.
+///   Examples in other type checkers are mypy's MYPYPATH environment variable,
+///   or pyright's stubPath configuration setting.
+/// - `workspace_root` is the root of the workspace,
+///   used for finding first-party modules
+/// - `site-packages` is the path to the user's `site-packages` directory,
+///   where third-party packages from ``PyPI`` are installed
+/// - `custom_typeshed` is a path to standard-library typeshed stubs.
+///   Currently this has to be a directory that exists on disk.
+///   (TODO: fall back to vendored stubs if no custom directory is provided.)
+impl ResolvedSearchPathOrder {
+    pub fn new(
+        extra_paths: Vec<PathBuf>,
+        workspace_root: PathBuf,
+        site_packages: Option<PathBuf>,
+        custom_typeshed: Option<PathBuf>,
+    ) -> Self {
+        Self(
+            extra_paths
+                .into_iter()
+                .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::Extra))
+                .chain(std::iter::once(ModuleSearchPath::new(
+                    workspace_root,
+                    ModuleSearchPathKind::FirstParty,
+                )))
+                // TODO fallback to vendored typeshed stubs if no custom typeshed directory is provided by the user
+                .chain(
+                    custom_typeshed.into_iter().map(|path| {
+                        ModuleSearchPath::new(path, ModuleSearchPathKind::StandardLibrary)
+                    }),
+                )
+                .chain(site_packages.into_iter().map(|path| {
+                    ModuleSearchPath::new(path, ModuleSearchPathKind::SitePackagesThirdParty)
+                }))
+                // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
+                .collect(),
```

Comment:
Hmm, to me it seems a fair bit more verbose but no more readable :/

```diff
--- a/crates/red_knot/src/module.rs
+++ b/crates/red_knot/src/module.rs
@@ -430,26 +430,40 @@ impl ResolvedSearchPathOrder {
         site_packages: Option<PathBuf>,
         custom_typeshed: Option<PathBuf>,
     ) -> Self {
-        Self(
+        let mut paths: Vec<ModuleSearchPath> = Vec::with_capacity(
+            extra_paths.len()
+                + 1
+                + usize::from(site_packages.is_some())
+                + usize::from(custom_typeshed.is_some()),
+        );
+
+        paths.extend(
             extra_paths
                 .into_iter()
-                .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::Extra))
-                .chain(std::iter::once(ModuleSearchPath::new(
-                    workspace_root,
-                    ModuleSearchPathKind::FirstParty,
-                )))
-                // TODO fallback to vendored typeshed stubs if no custom typeshed directory is provided by the user
-                .chain(
-                    custom_typeshed.into_iter().map(|path| {
-                        ModuleSearchPath::new(path, ModuleSearchPathKind::StandardLibrary)
-                    }),
-                )
-                .chain(site_packages.into_iter().map(|path| {
-                    ModuleSearchPath::new(path, ModuleSearchPathKind::SitePackagesThirdParty)
-                }))
-                // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
-                .collect(),
-        )
+                .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::Extra)),
+        );
+
+        paths.push(ModuleSearchPath::new(
+            workspace_root,
+            ModuleSearchPathKind::FirstParty,
+        ));
+
+        // TODO fallback to vendored typeshed stubs if no custom typeshed directory is provided by the user
+        paths.extend(
+            custom_typeshed
+                .into_iter()
+                .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::StandardLibrary)),
+        );
+
+        paths.extend(
+            site_packages.into_iter().map(|path| {
+                ModuleSearchPath::new(path, ModuleSearchPathKind::SitePackagesThirdParty)
+            }),
+        );
+
+        // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
+
+        Self(paths)
     }

```

Reply:
Yeah, I probably would forgo the `with_capacity` optimisation which you also won't get with `std::iter::chain` because `ExactSizeIterator` isn't implemented for chained (because it could overflow).

Reply:
> which you also won't get with `std::iter::chain` because `ExactSizeIterator` isn't implemented for chained (because it could overflow)

Ahh, I didn't realise that! But it makes sense now that you mention it.

I still kinda like the `chain()` personally, though 🤷

Reply:
I guess we could also `shrink_to_fit` the paths before returning, but probably doesn't matter given that few of these are ever created.